Refactor dotnet code gen

clrudolphi opened the pull request:
### 🤔 What's changed?

Rewrote the Dotnet code generation script to adopt the ruby erb gem approach used for other languages.
Modified the target .NET Framework settings by eliminating the use of .NET45 - targeting .NETStandard2.0 only.

Used a DevContainer so you'll see a devcontainer.json in the file set.

### ⚡️ What's your motivation? 

To bring the .NET implementation to parity with the java version.

### 🏷️ What kind of change is this?


- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### ♻️ Anything particular you want feedback on?

Per the question I posted on issue #230, does the ordering of properties of a entity matter when serialized? The serialization I have does not match the sequence of properties shown in the CCK examples.

I've used a Microsoft provided devcontainer image. Do you mind? Or should that be changed to the github image?
### 📋 Checklist:


- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [ ] I've changed the behaviour of the code
  - [X ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [X ] Users should know about my change
  - [X ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

mpkorstanje commented on dotnet/Makefile:
```diff
@@ -1,18 +1,24 @@
 
-clean:
-	rm -f Cucumber.Messages/Messages.cs .fixprotoc
+generate: require .generate-messages ## Generate c# code based on the schemas found in ../jsonschema and using the scripts in ../jsonschema/scripts for the generation
```
Now that dotnet is supported, you can also add it to the supported languages in `Makefile` in the root of the project.

mpkorstanje commented on dotnet/Cucumber.Messages.Specs/MessagesSpec.cs:
```diff
@@ -1,31 +1,68 @@
+            var jsonOptions = new JsonSerializerOptions();
+            jsonOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
+            jsonOptions.Converters.Add(new CustomEnumConverter<StepDefinitionPatternType>());
+            jsonOptions.DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull;
```
These json options are non trivial and will have to be replicated by anyone who wants to serialize messages. So I think it would be prudent to provide a serializer and deserializer class as part of the messages library. This should hopefully prevent any mistakes from taking hold.

You may also want to look at some of the other utility classes that come with messages. 

https://github.com/cucumber/messages/tree/main/java/src/main/java/io/cucumber/messages

When comparing do note that unlike many other languages, Java does not have an out of the box json serializer  On the other hand, the libraries Java does have update quite frequently. To avoid this cascading down from messages, the `MessageToNdjsonWriter` is constructed with a `Serializer` interface instead. This allows us to keep the serializer library at the edge of the dependency tree.

clrudolphi replied:
Thanks for the quick feedback. I'll take a look at whether the utility classes make sense in the dotnet ecosystem. 
I'll double check with Gaspar about the .net version (5 v 8). 
And I'll port over the Java test suite (or something similar).

clrudolphi replied:
Completed. Please take a look at what was added to the support classes to make serialization easier. It aligns with how the Java implementation works.

mpkorstanje reviewed (changes requested):
Looks good over all.

But I'm a bit puzzled as to why the CI hasn't been activated. I would have expected some build failures. I'll look into that.

For the mean time, I left a few comments. The first two are for myself. The third one you may want to address.

mpkorstanje commented:
> Per the question I posted on issue https://github.com/cucumber/messages/issues/230, does the ordering of properties of a entity matter when serialized? The serialization I have does not match the sequence of properties shown in the CCK examples.

Answered: https://github.com/cucumber/messages/issues/230#issuecomment-2197433270. The order does not matter.

> I've used a Microsoft provided devcontainer image. Do you mind? Or should that be changed to the github image?

As long as that choice falls within the common practices in the .Net ecosystem that is okay. The goal is to make sure the next contributor familiar with .Net understands what is going on.

I also don't think the Github provided containers are available on Dockerhub.

clrudolphi commented:
Updated the PR to include a ChangeLog entry.
Revised the code gen script to house all the generated C# classes in a /generated folder.
Added support classes in alignment with what the Java implementation has (for serialization support and TimeStamp conversions).
Added tests for the support classes (again, similar in nature to what the Java implementation does)
Dropped target framework support for .NET Framework 4.x (leaving .NET Standard 2.0 as the target).

clrudolphi commented:
@mpkorstanje please add @gasparnagy as a reviewer of this PR. Thanks

mpkorstanje commented on .github/workflows/test-dotnet.yml:
```diff
@@ -22,5 +22,6 @@ jobs:
         run: |
-          make
+         echo "No tests yet"
+         exit 1
```
Please add something sensible here.

clrudolphi replied:
i've modified that to invoke dotnet test

mpkorstanje commented on dotnet/Cucumber.Messages.Specs/ProtocolVersionTest.cs:
```diff
@@ -0,0 +1,19 @@
+        [Fact]
+        public void ProtocolVersionIsCorrect()
+        {
+            Assert.Matches(@"\d+\.\d+\.\d+(-SNAPSHOT)?", Cucumber.Messages.ProtocolVersion.Version);
```
Does .Net use snapshot versions?

clrudolphi replied:
It is not a community standard, no. Version numbering conventions are team-specific. Most teams do use some kind of suffix to designate pre release builds. 

I blindly copied what the Java implementation was doing. Sorry. I should have investigated what the prior .Net implementation had done and carried that forward. 

I'll revise as needed.

clrudolphi replied:
The prior version also used a suffix called SNAPSHOT; or at least that's what I can make of it (sorry, not well versed with github workflows). Open to changing that to whatever makes sense. I will ask Gaspar about it on Monday.

mpkorstanje replied:
No worries. I think that's a bit of copy paste from Gherkin. 

You can leave the `.csproj` file as is for now. I'll make a separate item to clean it up. It also means we need to change our release scripts that match against `VersionNumber` instead of `Version`.

https://github.com/cucumber/polyglot-release/blob/75e8af2c3398433d7a01862a863d038aa7d22dcb/polyglot-release#L70

clrudolphi replied:
Generalized that regex to look for an optional dot separated set of alphanumeric identifiers per the Semver standard as commonly used in .net apps.

gasparnagy commented on dotnet/Cucumber.Messages.Specs/Cucumber.Messages.Specs.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFramework>net5.0</TargetFramework>
+    <TargetFramework>net8.0</TargetFramework>
```
This is just our internal test project, not a deliverable. So this will not cause any issues. (But the other changes will be breaking anyway for .NET, I guess).

gasparnagy commented on jsonschema/scripts/codegen.rb:
```diff
@@ -214,6 +214,33 @@ def format_description(raw_description, indent_string: "")
+class Dotnet < Codegen
+  def initialize(paths)
+    language_type_by_schema_type = {
+      'integer' => 'long',
```
Are we sure that the 'integer' has to be a long in our case? Wouldn't `int` be enough?

@mpkorstanje Is there an explicit reason for requiring a 8 byte number for the messages? As far as I understand, we use numbers for line can column positions, where the maximum of the 4-byte int 2,147,483,647, should be more than sufficient. Even Visual Studio [treats line numbers as 4 byte `int`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.itextsnapshotline.linenumber?view=visualstudiosdk-2022#microsoft-visualstudio-text-itextsnapshotline-linenumber). Working with `long` is cumbersome in .NET because all other API uses `int` and converting them back and force is painful.

So my suggestion would be to use `int` here.

clrudolphi replied:
The same script generates the TimeStamp class which uses two longs to represent the number of seconds that have passed since 1 Jan 1970 and the number of partial seconds expressed as nanoseconds.
Using a signed int would work, but only for the next 13.5 years before the seconds field would overflow.
Would you all prefer that the expressive range of implementations be consistent? 
Would you prefer that we use Int32 for Location fields but Int64 for timestamp? (would require some fiddling with the script but I think do-able).

mpkorstanje replied:
I reckon the correct way of solving this would be to add the [`minimum` and maximum` keywords](https://json-schema.org/understanding-json-schema/reference/numeric#range) to the line and column fields. Then the code generation can determine the most appropriate type based on that.

clrudolphi replied:
@gasparnagy Is it worth waiting for this change to the schema to move through the ecosystem? I'll do what ever you two think best.

gasparnagy replied:
No. Let's just keep them as `long`. It does not worth the effort.

gasparnagy reviewed (changes requested):
Thx for the code, it looks good. 

I added my notes. Most of them minor, there are only a few tricky questions.

clrudolphi commented on dotnet/Cucumber.Messages/generated/Attachment.cs:
```diff
@@ -0,0 +1,157 @@
+        int hash = 17;
+        if (Body != null)
+          hash = hash * 31 + Body.GetHashCode();
+        if (ContentEncoding != null)
```
@gasparnagy Any concerns with the rather rudimentary hash algorithm I'm using here?

gasparnagy replied:
No. I'm fine with that.

clrudolphi commented on dotnet/Cucumber.Messages/Cucumber.Messages.csproj:
```diff
@@ -18,8 +19,8 @@
-    <Copyright>Copyright &#xA9; Cucumber Ltd, TechTalk</Copyright>
-    <Description>Protocol Buffer messages for Cucumber's and SpecFlow's inter-process communication</Description>
+    <Copyright>Copyright © Cucumber Ltd, TechTalk</Copyright>
+    <Description>Cucumber messages for Cucumber's and SpecFlow's inter-process communication</Description>
```
Adopted your wording.

clrudolphi commented on dotnet/Cucumber.Messages/Cucumber.Messages.csproj:
```diff
@@ -32,21 +33,11 @@
-    </PackageReference>
-
-    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.3">
+     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.3">
```
Removed

clrudolphi commented on dotnet/Cucumber.Messages/ProtocolVersion.cs:
```diff
@@ -0,0 +1,18 @@
+        {
+            get
+            {
+                return Assembly.GetExecutingAssembly().GetName().Version.ToString();
```
Adopted your suggested code into that function.

clrudolphi commented on dotnet/Cucumber.Messages/generated/Attachment.cs:
```diff
@@ -0,0 +1,157 @@
+ * It is not to be used for runtime errors raised/thrown during execution. This
+ * is captured in `TestResult`.
+ */
+// Generated code
```
done

luke-hill commented:
Dependent on the timeframes involved this will maybe require a large conflict resolution as I'm in the process of refactoring the codegen (Only a small refactor, but lots of moving of files around), here: https://github.com/cucumber/messages/pull/228

I'm currently at a bit of an impasse in terms of contributing for the coming weeks, so if this is merged in in the next 1-2 weeks you'll have no issues and I'll fix up conflicts my end. But just letting you know :+1:

gasparnagy commented on dotnet/Cucumber.Messages/Cucumber.Messages.csproj:
```diff
@@ -32,21 +33,11 @@
-  <ItemGroup>
-    <Compile Remove="Messages.cs"/>
-    <Protobuf Include="../messages.proto" OutputDir="$(MsBuildThisFileDirectory)" GrpcServices="None"/>
+     <PackageReference Include="System.Text.Json" Version="8.0.3" />
```
I see. It seems that Reqnroll depends on Microsoft.Extensions.DependencyModel v6.0.0 anyways. And that version depends on System.text.json v6.0.0. So we can safely depend on that version without changing the overall dependency chain. Would system.text.json v6.0.0 be enough for our needs?

clrudolphi replied:
Yes, that will work. Updated PR.

gasparnagy commented on dotnet/Cucumber.Messages/NdjsonSerializer.cs:
```diff
@@ -3,23 +3,30 @@
-    public class NdjsonSerializer
+    public static class NdjsonSerializer
     {
+        private static object lockObject = new object();
```
This solution is probably good enough, but for such thread-safe initializations, I generally prefer using the `Lazy<T>` class, because the lock object version needs to aquire a lock for every single property get, even after the initialization period.

clrudolphi replied:
Ah. I had completely forgotten about the option to use Lazy. (I'm too old school for such new sparkly things) 
I'll change it and resubmit.

clrudolphi replied:
Done. Revised the PR to make this use Lazy<T>

mpkorstanje commented:
@luke-hill cheers. 

@clrudolphi @gasparnagy take your time, I won't be looking at opensource this week until the end of the weekend.

mpkorstanje commented on dotnet/Cucumber.Messages/CucumberMessagEnumConverter.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Cucumber.Messages
+{
+    public class CucumberMessagEnumConverter<T> : JsonConverter<T> where T : struct, Enum
```
Versioning is easier if a library only exposes the essentials. Consider making this and any other classes not explicitly intended to be part of the API package private (or the .Net) equivalent.

clrudolphi replied:
Agreed. I'll make this class internal to the Cucumber.Messages assembly.

mpkorstanje commented on dotnet/Cucumber.Messages/NdjsonSerializer.cs:
```diff
@@ -0,0 +1,53 @@
+            return NdjsonSerializer.Deserialize<Envelope>(json);
+        }
+
+        public static T Deserialize<T>(string json)
```
I would only ever expect users of this API to read and write envelopes. So `Serialize<T>` and `Deserialize<T>` can probably be package private.

clrudolphi replied:
I had left it public in order to test it (via elements more granular than Envelopes). Now that we're over that hurdle, I'll remove the single such test and make these private.

clrudolphi commented:
I've also added a line to MessagetoNdjsonReader and Writer classes to Close their internally held streamwriter upon Dispose().

luke-hill commented on dotnet/.github/dependabot.yml:
```diff
@@ -0,0 +1,12 @@
+ - package-ecosystem: "devcontainers"
+   directory: "/"
+   schedule:
+     interval: weekly
```
Is this still needed?

luke-hill commented on dotnet/Cucumber.Messages.Specs/BasicMessageSerializationTests.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            var stepDefinitionNDJSON = @"{""stepDefinition"":{""id"":""0"",""pattern"":{""source"":""I have {int} cukes in my belly"",""type"":""CUCUMBER_EXPRESSION""},""sourceReference"":{""location"":{""line"":3},""uri"":""samples/minimal/minimal.feature.ts""}}}
+";          
+            var soureReference = new SourceReference("samples/minimal/minimal.feature.ts",
```
sourceReference*

clrudolphi replied:
Fixed misspelled variable name.

Re: the comment above about "Is this still needed?" - does that refer to the dependabot itself or the use of the devcontainer?

luke-hill replied:
The use of dependabot. As I know we were discussing using renovate which is used most other places.

luke-hill commented on dotnet/Cucumber.Messages.Specs/MessagesTest.cs:
```diff
@@ -0,0 +1,27 @@
+        public void IsValidWhenNoRequiredFieldsAreMissing()
+        {
+            // This should succeed as no constructor arguments to an Envelope are required.
+            var env = new Envelope(null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null);
```
Should this have an assertion like no error is raised?

clrudolphi replied:
I was duplicating the approach taken by the Java implementation. It demonstrates that no arguments are required to the main constructor of an Envelope.
The alternative here, IMO, would be to remove the test entirely.

mpkorstanje replied:
There is an `assertDoesNotThrow(() -> { some code })` in Java, but its kinda redundant in single line test case.

luke-hill commented on dotnet/Cucumber.Messages.Specs/MessagesTest.cs:
```diff
@@ -0,0 +1,27 @@
+        [Fact]
+        public void ThrowsWhenRequiredFieldsAreMissing()
+        {
+            Assert.Throws<ArgumentNullException>(() => new Background( null, null, null, null, null, null));
```
Not sure if you care about linting but space before `(`

clrudolphi replied:
Corrected.

luke-hill commented on dotnet/Cucumber.Messages.Specs/NdjsonStreamSerializationTests.cs:
```diff
@@ -0,0 +1,91 @@
+        public void IgnoresEmptyLines()
+        {
+            MemoryStream memoryStream = new MemoryStream(Encoding.UTF8.GetBytes("{}\n{}\n\n{}\n"));
+            var enumerator= new NdjsonMessageReader(memoryStream).GetEnumerator();
```
`enumerator = `

clrudolphi replied:
Corrected.

luke-hill commented on dotnet/Cucumber.Messages/Cucumber.Messages.csproj:
```diff
@@ -18,9 +19,9 @@
-    <Description>Protocol Buffer messages for Cucumber's and SpecFlow's inter-process communication</Description>
-    <PackageTags>cucumber-messages cucumber specflow gherkin</PackageTags>
+    <Copyright>Copyright © Cucumber Ltd, TechTalk</Copyright>
+    <Description>Cucumber messages to represent source data and execution results in Cucumber-family tools</Description>
```
> JSON schema-based messages for Cucumber's inter-process communication

clrudolphi replied:
Adopted your wording.

luke-hill commented on dotnet/Cucumber.Messages/NdjsonSerializer.cs:
```diff
@@ -0,0 +1,51 @@
+    {
+        private static Lazy<JsonSerializerOptions> _jsonOptions = new Lazy<JsonSerializerOptions>(() =>
+        {
+            var jso = new JsonSerializerOptions();
```
I'm guessing this one is meant to be jso, is there maybe a better name for this to avoid confusion over a misname? `options` perhaps?

clrudolphi replied:
Adopted the use of the variable name 'options'.

luke-hill commented on jsonschema/scripts/codegen.rb:
```diff
@@ -214,6 +214,33 @@ def format_description(raw_description, indent_string: "")
+    "List<#{type_name}>"
+  end
+
+  def format_description(raw_description, indent_string: "")
```
For now I've just got a couple of fixes to suggest here as the entire format is changing. But if you change this to single quotes and the item suggested below then that should be fine for now. Until the stuff gets merged

clrudolphi replied:
indent_string is now ''

luke-hill commented on jsonschema/scripts/codegen.rb:
```diff
@@ -214,6 +214,33 @@ def format_description(raw_description, indent_string: "")
+
+    raw_description
+      .split("\n")
+      .map { |line| line.strip() }
```
.strip does not require a `()` argument here. Nor in the one below

clrudolphi replied:
adopted

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+/**
+ * Represents the <%= class_name(key) %> message in Cucumber's message protocol
+ * @see <a href="https://github.com/cucumber/messages" >Github - Cucumber - Messages</a>
+ <%- if not (schema['description'] || []).empty? -%>
```
`%- unless (schema['description'] || []).empty? -%>`

clrudolphi replied:
adopted

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+{
+  <%- schema['properties'].each do |property_name, property| 
+      required = (schema['required'] || []).index(property_name) 
+      isValueType = type_for(class_name(key), property_name, property) == "long" || type_for(class_name(key), property_name, property) == "bool"
```
this logic feels quite complex and can mess with the ordering of the lines. Is there any way we can push this into the class proper?

clrudolphi replied:
This cannot be pushed into the .NET class. The intent is to identify those properties and fields that are optional value types. The declarations for these need to be different than a regular long or bool; they are declared as Nullable<T> (see lines 35/37 and 65/67 for a comparison). 
I don't understand your meaning of "can mess with the ordering of the lines" - what is the impact? Is there an example within the code to demonstrate?

luke-hill replied:
Sorry I meant more migrating this into the `Dotnet` ruby class that would determine this logic so we can make the overall erb template smaller. But I think this is probably best done after this has been merged and then I tag you for review. It will probably make more sense that way and you've done a lot of work here I don't want to delay / overcomplicate things.

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+      required = (schema['required'] || []).index(property_name) 
+      isValueType = type_for(class_name(key), property_name, property) == "long" || type_for(class_name(key), property_name, property) == "bool"
+  -%>
+    <%- if not (property['description'] || []).empty? -%>
```
`<%- unless (property['description'] || []).empty? -%>`

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+  <%- end -%>
+
+  <%- if (schema['required'] || []).empty? -%>
+  <%- schema['properties'].each_with_index do |(property_name, property), index| -%>
```
this can become `.each` and you can omit the index parameter in the pipes

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+        return new <%= class_name(key) %>(
+        <%- schema['properties'].each_with_index do |(property_name_2, property_2), index_2| -%>
+          <%- if property_name_2 == property_name -%>
+            Require<<%= type_for(class_name(key), property_name, property) -%>>(<%= property_name %>, "<%= capitalize(property_name) %>", "<%= class_name(key) %>.<%= capitalize(property_name) %> cannot be null")<%= index_2 < schema['properties'].length-1 ? ',' : '' %>
```
```suggestion
            Require<<%= type_for(class_name(key), property_name, property) -%>>(<%= property_name %>, "<%= capitalize(property_name) %>", "<%= class_name(key) %>.<%= capitalize(property_name) %> cannot be null")<%= index < schema['properties'].length - 1 ? ',' : '' %>
```

clrudolphi replied:
adopted

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+    public static <%= class_name(key) %> Create(<%= type_for(class_name(key), property_name, property) -%> <%= property_name %>) 
+    {
+        return new <%= class_name(key) %>(
+        <%- schema['properties'].each_with_index do |(property_name_2, property_2), index_2| -%>
```
```suggestion
        <%- schema['properties'].each_with_index do |(property_name_2, property_2), index| -%>
```

clrudolphi replied:
adopted

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+          <%- if property_name_2 == property_name -%>
+            Require<<%= type_for(class_name(key), property_name, property) -%>>(<%= property_name %>, "<%= capitalize(property_name) %>", "<%= class_name(key) %>.<%= capitalize(property_name) %> cannot be null")<%= index_2 < schema['properties'].length-1 ? ',' : '' %>
+          <%- else -%>
+            null<%= index_2 < schema['properties'].length-1 ? ',' : '' %>
```
```suggestion
            null<%= index < schema['properties'].length - 1 ? ',' : '' %>
```

clrudolphi replied:
adopted

luke-hill commented on jsonschema/scripts/templates/dotnet.dotnet.erb:
```diff
@@ -0,0 +1,150 @@
+
+    public <%= class_name(key) %>(
+    <%- schema['properties'].each_with_index do |(property_name, property), index| 
+        isValueType = type_for(class_name(key), property_name, property) == "long" || type_for(class_name(key), property_name, property) == "bool"
```
If this is pushed into your class you then get the win here

luke-hill replied:
Ignore this comment @clrudolphi based on the above comment. We will tackle this after the fact once this is approved and merged.

luke-hill reviewed (commented):
Popped some linting suggestions. I've tried to keep them brief / relevant. 

I think it's probably best we get this merged in first and then merge my codegen stuff in after as there's still a lot of stuff to tidy up but probably best I tackle it.

gasparnagy commented on dotnet/Cucumber.Messages/CucumberMessagEnumConverter.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Cucumber.Messages
+{
+    internal class CucumberMessagEnumConverter<T> : JsonConverter<T> where T : struct, Enum
```
Just found a missing "e". CucumberMessag**e**EnumConverter

clrudolphi replied:
Corrected.

clrudolphi commented:
Thanks for your review comments so far. 
I think the only outstanding item is the comment from @luke-hill about the dotnet generation script and the complexity required when declaring Nullable<T> properties/fields.

clrudolphi commented:
@mpkorstanje @luke-hill Any chance to close this out today or tomorrow? I'm going on holiday until the 1st of August and would be good to finish this so that @luke-hill can proceed with his other refactoring.

luke-hill commented:
@clrudolphi consider all my comments that have not been addressed to be "deferable" I don't want to hold anything up.

Just consider the actual dotnet stuff from Gaspar / Rien now. I can do some more fine tuning after your stuff is merged in a separate PR entirely to both this one and my companion one. Which would probably make things easier for you to review also.

mpkorstanje reviewed (approved):
Looks good! 

@gasparnagy any remarks? If not, feel free to merge.

gasparnagy reviewed (approved):
Shit it 😀!

clrudolphi commented:
> Shit it 😀!

Funniest merge approval comment ever!

gasparnagy commented:
😀 Autocorrect... But you know what I meant.

mpkorstanje commented:
@gasparnagy have you made a release before?

gasparnagy commented:
No, I don't think so.

luke-hill commented:
Just posting here. Let's hold off releasing until I get my codegen stuff in also. As then it's a chunky enough release for a major.

mpkorstanje commented:
No problem.
